### PR TITLE
Implement ImageData constructors #15671 -- 1

### PR DESCRIPTION
**Pull request by tnielens**
Implement ImageData constructors
* new constructors in the imagedata.rs file
* modification of the ImageData::new method
* new method ImageData::from_vec
* new wpt test file imagedata.html 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes affect #15671  (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15821)

### CODE REVIEW

File: components/script/dom/imagedata.rs [hunk 1]
```diff
@@ -27,31 +28,45 @@ pub struct ImageData {
...
+            width: width,
+            height: height,
+            data: Heap::default(),
+        };
 
-        reflect_dom_object(imagedata,
-                           global, ImageDataBinding::Wrap)
+        let len = width * height * 4;
+        match data {
+            Some(jsobject) => (*imagedata).data.set(jsobject),
```
Comment (nox) [inline components/script/dom/imagedata.rs]: This needs various checks, such as whether the given object is indeed a proper typed array.

File: components/script/dom/imagedata.rs [hunk 2]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
             vec
         }
     }
 
     pub fn get_size(&self) -> Size2D<i32> {
         Size2D::new(self.Width() as i32, self.Height() as i32)
     }
-}
 
+    pub fn Constructor(global: &GlobalScope, width: u32, height: u32) -> Fallible<Root<Self>> {
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Nit: spec link.

File: components/script/dom/imagedata.rs [hunk 3]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+        if width == 0 || height == 0 {
+            Err(Error::IndexSize)
+        } else {
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Nit: remove the `else` and short-circuit in the error case by returning early.

File: components/script/dom/imagedata.rs [hunk 4]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+            Ok(Self::new(global, width, height, None))
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn Constructor_(cx: *mut JSContext,
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Nit: spec link.

File: components/script/dom/imagedata.rs [hunk 5]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+                        width: u32,
+                        opt_height: Option<u32>)
+                        -> Fallible<Root<Self>> {
+        typedarray!(in(cx) let data_res: Uint8ClampedArray = data);
+        let mut array = match data_res {
+            Ok(data) => data,
+            Err(_)   => {
+                return Err(Error::Type("Argument to Image data is not an ArrayBufferView".to_owned()));
+            }
+        };
```
Comment (nox) [inline components/script/dom/imagedata.rs]:
```rust
let mut array = data_res.map_err(|_| {
    Error::Type("Argument to ImageData is not an ArrayBufferView".to_owned())
});
```

File: components/script/dom/imagedata.rs [hunk 6]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+        let mut len;
+        unsafe {
+            len = array.as_slice().len() as u32;
+        }
```
Comment (nox) [inline components/script/dom/imagedata.rs]:
```rust
let byte_len = unsafe { array.as_slice().len() as u32 };
```

File: components/script/dom/imagedata.rs [hunk 7]
```diff
@@ -27,31 +28,45 @@ pub struct ImageData {
...
             };
-            Uint8ClampedArray::create(cx, data, js_object.handle_mut()).unwrap();
-            (*imagedata).data.set(js_object.get());
+            Uint8ClampedArray::create(cx, data, typedarray.handle_mut()).unwrap();
+            Self::new(global, width, height, Some(typedarray.get()))
         }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn new(global: &GlobalScope, width: u32, height: u32, data: Option<*mut JSObject>) -> Root<ImageData> {
```
Comment (nox) [inline components/script/dom/imagedata.rs]: The whole method should be unsafe, and please rename with to `new_with_jsobject`.

File: components/script/dom/imagedata.rs [hunk 8]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+                return Err(Error::Type("Argument to Image data is not an ArrayBufferView".to_owned()));
+            }
+        };
+        let mut len;
+        unsafe {
+            len = array.as_slice().len() as u32;
+        }
+        if len % 4 != 0 {
+            Err(Error::InvalidState)
+        } else {
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Nit: remove the `else` and short-circuit in the error case by returning early.

File: components/script/dom/imagedata.rs [hunk 9]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+            len /= 4;
```
Comment (nox) [inline components/script/dom/imagedata.rs]:
```rust
let len = byte_len / 4;
```

File: components/script/dom/imagedata.rs [hunk 10]
```diff
@@ -60,17 +75,57 @@ impl ImageData {
...
+            let height = len / width;
+            if len % width != 0 {
+                Err(Error::IndexSize)
+            } else if opt_height.is_some() && height != opt_height.unwrap() {
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Nit: remove the `else` and short-circuit in the error case by returning early.

Comment (nox) [inline components/script/dom/imagedata.rs]:
```rust
if opt_height.map_or(false, |x| x != height) {
    return Err(Error::IndexSize);
}
```

File: components/script/dom/webidls/ImageData.webidl [hunk 11]
```diff
@@ -9,9 +9,10 @@
  * You are granted a license to use, reproduce and create derivative works of this document.
  */
 
-//[Constructor(unsigned long sw, unsigned long sh),
-//Constructor(Uint8ClampedArray data, unsigned long sw, optional unsigned long sh),
-[Exposed=(Window,Worker)]
+[Constructor(unsigned long sw, unsigned long sh),
+// Constructor(Uint8Array data, unsigned long sw, optional unsigned long sh),
+ Constructor(object data, unsigned long sw, optional unsigned long sh),
```
Comment (nox) [inline components/script/dom/webidls/ImageData.webidl]: Nit: remove the commented-out line and instead write the first argument as:

```
/*Uint8Array*/ object data
```

**Review by nox — CHANGES_REQUESTED** [review]
I haven't yet reviewed whether this follows the spec, nor the tests.

File: components/script/dom/imagedata.rs (hunk 1 above)
Comment (montrivo) [inline components/script/dom/imagedata.rs]: Should I repeat various checks already done in `Constructor` and `Constructor_` or should I remove the `pub` modifier and comment arguments preconditions?

File: components/script/dom/imagedata.rs [hunk 12]
```diff
@@ -27,31 +28,45 @@ pub struct ImageData {
...
-    pub fn new(global: &GlobalScope, width: u32, height: u32, mut data: Option<Vec<u8>>) -> Root<ImageData> {
-        let imagedata = box ImageData {
-            reflector_: Reflector::new(),
-            width: width,
-            height: height,
-            data: Heap::default(),
-        };
-        let len = width * height * 4;
-
+    pub fn from_vec(global: &GlobalScope, width: u32, height: u32, mut data: Option<Vec<u8>>) -> Root<ImageData> {
```
Comment (montrivo) [inline components/script/dom/imagedata.rs]: The `new` method doesn't validate arguments, neither returns a `Faillible<Root<ImageData>>`. Should it do the same kind of validations as the two new constructors ?

**Comment by nox** [comment]
Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/imagedata.rs, line 31 at r1](https://reviewable.io:443/reviews/servo/servo/15821#-KeQdu7pqa6UUmV7orPn-r1-31:-KebwgWNAI59YZ4d4oj7:b-lan7r1) ([raw file](https://github.com/servo/servo/blob/b065bc872ddc2709f8bdac7e3a387070f61f15e5/components/script/dom/imagedata.rs#L31)):*
<details><summary><i>Previously, montrivo wrote…</i></summary><blockquote>

The `new` method doesn't validate arguments, neither returns a `Faillible<Root<ImageData>>`. Should it do the same kind of validations as the two new constructors ?
</blockquote></details>

You can reuse `new_with_jsobject`, but the whole method should panic in case of an error.

---

*[components/script/dom/imagedata.rs, line 59 at r1](https://reviewable.io:443/reviews/servo/servo/15821#-KeQdu7pqa6UUmV7orPn-r1-59:-KebwSu2p7ceaXgwJ_SI:bif9tvl) ([raw file](https://github.com/servo/servo/blob/b065bc872ddc2709f8bdac7e3a387070f61f15e5/components/script/dom/imagedata.rs#L59)):*
<details><summary><i>Previously, montrivo wrote…</i></summary><blockquote>

Should I repeat various checks already done in `Constructor` and `Constructor_` or should I remove the `pub` modifier and comment arguments preconditions?
</blockquote></details>

Make this method return `Fallible<Root<ImageData>>` and reuse that in every other method.

---

*[components/script/dom/imagedata.rs, line 109 at r1](https://reviewable.io:443/reviews/servo/servo/15821#-KeQdu7pqa6UUmV7orPn-r1-109:-KebwFo9-B_dsjDqDl2i:bwyxmfa) ([raw file](https://github.com/servo/servo/blob/b065bc872ddc2709f8bdac7e3a387070f61f15e5/components/script/dom/imagedata.rs#L109)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary><blockquote>

```rust
let mut array = data_res.map_err(|_| {
    Error::Type("Argument to ImageData is not an ArrayBufferView".to_owned())
});
```
</blockquote></details>

You didn't address this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15821)*

**Comment by nox** [comment]
-S-awaiting-review 

Seems ok! Running try because I think there will be test expectations to change.

@bors-servo try

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15821#-:-Kehm7gcEfHeAYR0RVay:bbkxame)*

**Comment by nox** [comment]
```
diff --git a/tests/wpt/metadata/MANIFEST.json b/tests/wpt/metadata/MANIFEST.json
index 6c2c84a..c1b47c4 100644
--- a/tests/wpt/metadata/MANIFEST.json
+++ b/tests/wpt/metadata/MANIFEST.json
@@ -175569,7 +175569,7 @@
    "testharness"
   ],
   "html/semantics/embedded-content/the-canvas-element/imagedata.html": [
-   "b0dade5b5f0a3e8b7798f52dbcb2ae0f23fed150",
+   "04c670c9a045d924725bedfb09be665f73c3a0ad",
    "testharness"
   ],
   "html/semantics/embedded-content/the-canvas-element/initial.colour.html": [
```

I'll check whether the other build jobs finish successfully.

**Review by nox — APPROVED** [review]

**Comment by nox** [comment]
```
  ▶ Unexpected subtest result in /2dcontext/pixel-manipulation/2d.imageData.object.ctor.html:
  │ FAIL [expected PASS] ImageData does not have a usable constructor
  │   → assert_throws: function "function () { new window.ImageData(1,1); }" did not throw
  │ 
  │ @http://web-platform.test:8000/2dcontext/pixel-manipulation/2d.imageData.object.ctor.html:23:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  │ _addTest/</<@http://web-platform.test:8000/common/canvas-tests.js:62:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  └ _addTest/<@http://web-platform.test:8000/common/canvas-tests.js:59:9

  ▶ Unexpected subtest result in /html/dom/interfaces.worker.html:
  └ PASS [expected FAIL] ImageData interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] ImageData interface object length
```

The 2 `PASS` are good news, the `FAIL` isn't.

**Comment by nox** [comment]
The test `/2dcontext/pixel-manipulation/2d.imageData.object.ctor.html` doesn't pass in either Safari nor Firefox, it's probably outdated.

**Comment by nox** [comment]
@jdm Should it be marked as `FAIL`, or should we fix the test right now in this PR?

**Comment by nox** [comment]
@Ms2ger What do you think about the failing test? Should we just mark it as FAIL and report the erroneous test upstream?

**Comment by Ms2ger** [comment]
Mark it as fail; I want wpt CI to run on the change.

File: tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html [hunk 13]
```diff
@@ -0,0 +1,55 @@
+<!doctype html>
+<meta charset="utf-8">
+<title></title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+function constructor_domexceptions(domExcepName, description, func) {
+    test(function() {
+        var result;
+        try {
```
Comment (Ms2ger) [inline tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html]: Use `assert_throws`

File: tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html [hunk 14]
```diff
@@ -0,0 +1,68 @@
+<!doctype html>
+<meta charset="utf-8">
+<title></title>
```
Comment (nox) [inline tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html]: Add a title.

File: tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    })},
+    "ImagaData(w, h), height cannot be 0");
+
+test(function() {
+    var imageData = new ImageData(2, 3);
+    assert_equals(imageData.width, 2);
+    assert_equals(imageData.height, 3);
+    assert_equals(imageData.data.length, 24);
+    assert_true(imageData.data instanceof Uint8ClampedArray);
+}, "ImagaData(w, h), exposed attributes check");
```
Comment (nox) [inline tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html]: Please reindent all tests to look like this one.

File: tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+test(function(){
+    assert_throws(
+            "IndexSizeError",
+            function(){
+                new ImageData(0, 1);
+    })},
+    "ImagaData(w, h), width cannot be 0");
```
Comment (nox) [inline tests/wpt/web-platform-tests/html/semantics/embedded-content/the-canvas-element/imagedata.html]: Please fix all the spelling errors, it's `ImageData`.

**Review by nox — DISMISSED** [review]
A few cosmetic changes on the test and this will be ready.

**Review by Ms2ger — APPROVED** [review]

**Pull request merged**